src/doc/trpl/concurrency.md

```diff
@@ -0,0 +1,376 @@
+threadsafe, we wouldn't want to implement `Send`, and so the compiler will help
+us enforce that it can't leave the current thread.
+
+### `Sync`
+
+The second of these two trait is called [`Sync`](../std/marker/trait.Sync.html).
+When a type `T` implements `Sync`, it indicates to the compiler that something
+of this type has no possibility of introducing memory unsafety when passing
+`&T`s between threads.
+
+For example, sharing immutable data with an atomic reference count is
+threadsafe. Rust provides a type like this, `Arc<T>`, and it implements `Sync`,
+so that it could be safely shared between threads.
+
+These two traits allow you to use the type system to make strong guarantees
+about the properties of your code under concurrency. Before we demonstrate
+why, we need to learn how to create a concurrent Rust program in the first
+place!
+
+## Threads
+
+Rust's standard library provides a library for 'threads', which allow you to
+run Rust code in parallel. Here's a basic example of using `Thread`:
+
+```
+use std::thread::Thread;
+
+fn main() {
+    Thread::scoped(|| {
+        println!("Hello from a thread!");
+    });
+}
+```
+
+The `Thread::scoped()` method accepts a closure, which is executed in a new
+thread. It's called `scoped` because this thread returns a join guard:
+
+```
+use std::thread::Thread;
+
+fn main() {
+    let guard = Thread::scoped(|| {
+        println!("Hello from a thread!");
+    });
+
+    // guard goes out of scope here
+}
+```
+
+When `guard` goes out of scope, it will block execution until the thread is
+finished. If we didn't want this behaviour, we could use `Thread::spawn()`:
+
+```
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    Thread::spawn(|| {
+        println!("Hello from a thread!");
+    });
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+We need to `sleep` here because when `main()` ends, it kills all of the
+running threads.
+
+[`scoped`](std/thread/struct.Builder.html#method.scoped) has an interesting
+type signature:
+
+```text
+fn scoped<'a, T, F>(self, f: F) -> JoinGuard<'a, T>
+    where T: Send + 'a,
+          F: FnOnce() -> T,
+          F: Send + 'a
+```
+
+Specifically, `F`, the closure that we pass to execute in the new thread. It
+has two restrictions: It must be a `FnOnce` from `()` to `T`. Using `FnOnce`
+allows the clousre to take ownership of any data it mentions from the parent
+thread. The other restriction is that `F` must be `Send`. We aren't allowed to
+transfer this ownership unless the type thinks that's okay.
+
+Many languages have the ability to execute threads, but it's wildly unsafe.
+There are entire books about how to prevent errors that occur from shared
+mutable state. Rust helps out with its type system here as well, by preventing
+data races at compile time. Let's talk about how you actually share things
+between threads.
+
+## Safe Shared Mutable State
```

Comment:
I wonder if introducing shared state first, and then introducing mutability separately would be nice. It allows one to talk about `Arc` in a more focused way, since immutable shared data is a reasonable thing too.

---

Also, with https://github.com/rust-lang/rfcs/pull/458, there will be a better (when it works) way to shared data between threads than `Arc<T>`: just use references `&T`/`&mut T` and `Thread::scoped`. E.g. the code written here could be written as:

```rust
let guards = data.mut_iter()
     .map(|val| {
         Thread::scoped(move || { *val += 1 })
    })
    .collect::<Vec<JoinGuard<()>>>();
```

Of course, this isn't implemented yet, but I think this aspect of Rust's concurrency is one of the most compelling (perfectly safe pointers into another threads stack). I guess this is just a heads-up. :smile: